File: encode-and-decode-strings/Codec271.java
// https://leetcode.com/problems/encode-and-decode-strings/

import java.util.*;

class Codec271 {

    // T: O(n), S: O(n)
    public String encode(List<String> strs) {
        var sb = new StringBuilder();
        for (var str : strs) {
            int len = str.length();
            sb.append(len);
            sb.append("#");
            sb.append(str);
        }
        return sb.toString();
    }

    // T: O(n), S: O(n)
    public List<String> decode(String s) {
        var result = new ArrayList<String>();
        int len = s.length();
        int start = 0, end = 0;
        while (start < len) {
            while (s.charAt(end) != '#') {
                end++;
            }
            int wordLen = Integer.parseInt(s.substring(start, end));
            var word = s.substring(end + 1, end + wordLen + 1);
            result.add(word);
            start = end + wordLen + 1;
            end = start;
        }
        return result;
    }

    public static void main(String[] args) {
        Codec271 codec = new Codec271();
        List<String> strs = Arrays.asList("Hello", "World");
        String encoded = codec.encode(strs);
        System.out.println("Encoded: " + encoded);
        List<String> decoded = codec.decode(encoded);
        System.out.println("Decoded: " + decoded);
    }

}

File: encode-and-decode-strings/codec271.py
# https://leetcode.com/problems/encode-and-decode-strings/

from typing import List


class Codec271:

    # T: O(n), S: O(n)
    def encode(self, strs: List[str]) -> str:
        encoded_parts = []
        for s in strs:
            encoded_parts.append(f"{len(s)}#{s}")
        return ''.join(encoded_parts)

    # T: O(n), S: O(n)
    def decode(self, s: str) -> List[str]:
        result = []
        i = 0
        while i < len(s):
            j = i
            while s[j] != '#':
                j += 1
            length = int(s[i:j])
            result.append(s[j+1:j+1+length])
            i = j + 1 + length
        return result


def main() -> None:
    codec = Codec271()
    encoded = codec.encode(["Hello", "World"])
    print("Encoded:", encoded)
    decoded = codec.decode(encoded)
    print("Decoded:", decoded)


if __name__ == "__main__":
    main()
